In [2]:
import pandas as pd

data = pd.read_csv('/Users/tyco/Desktop/MADS/sanovo/data/raw/HSTMachineError.csv')
data.head(5)

MerId  MerType                  MerDate  MerNumber  MerSubNumber  \
0  11897185        7  2025-01-30 10:22:00.000         10            22   
1  11897186       21  2025-01-30 10:21:42.000        460            12   
2  11897187       23  2025-01-30 10:21:56.000        460            12   
3  11897188       21  2025-01-30 10:21:56.000        460            12   
4  11897189       23  2025-01-30 10:21:57.000        460            12   

   MerData1  MerData2  MerData3  MerData4  MerData5  MerMinId  MerText  \
0     181.0     180.0      90.0       NaN       NaN      4620      NaN   
1       1.0      22.0      41.0       NaN       NaN      4620      NaN   
2       0.0       0.0       0.0       NaN       NaN      4620      NaN   
3       1.0      22.0      41.0       NaN       NaN      4620      NaN   
4       0.0       0.0       0.0       NaN       NaN      4620      NaN   

   MerDeviceType  MerDeviceNumber  MerPriority  
0              0                0            0  
1              0                0            0  
2              0                0            0  
3              0                0            0  
4              0                0            0

In [3]:
# Create a copy of the data
data_copy = data.copy()

In [4]:
data_filtered = data_copy[(data_copy['MerType'].between(21, 24)) | (data_copy['MerType'] == 8)]
data_filtered['Error/warnings'] = data_filtered['MerType'].astype(str) + '_' + data_filtered['MerNumber'].astype(str) + '_' + data_filtered['MerData2'].astype(str) + '_' + data_filtered['MerData3'].astype(str) 
data_filtered.head(5)

/var/folders/pt/00xl4yzd64s0vbx9w9tzh94m0000gn/T/ipykernel_85212/1200778792.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['Error/warnings'] = data_filtered['MerType'].astype(str) + '_' + data_filtered['MerNumber'].astype(str) + '_' + data_filtered['MerData2'].astype(str) + '_' + data_filtered['MerData3'].astype(str)


MerId  MerType                  MerDate  MerNumber  MerSubNumber  \
1  11897186       21  2025-01-30 10:21:42.000        460            12   
2  11897187       23  2025-01-30 10:21:56.000        460            12   
3  11897188       21  2025-01-30 10:21:56.000        460            12   
4  11897189       23  2025-01-30 10:21:57.000        460            12   
5  11897190       21  2025-01-30 10:21:57.000        460            12   

   MerData1  MerData2  MerData3  MerData4  MerData5  MerMinId  MerText  \
1       1.0      22.0      41.0       NaN       NaN      4620      NaN   
2       0.0       0.0       0.0       NaN       NaN      4620      NaN   
3       1.0      22.0      41.0       NaN       NaN      4620      NaN   
4       0.0       0.0       0.0       NaN       NaN      4620      NaN   
5       1.0      22.0      41.0       NaN       NaN      4620      NaN   

   MerDeviceType  MerDeviceNumber  MerPriority    Error/warnings  
1              0                0            0  21_460_22.0_41.0  
2              0                0            0    23_460_0.0_0.0  
3              0                0            0  21_460_22.0_41.0  
4              0                0            0    23_460_0.0_0.0  
5              0                0            0  21_460_22.0_41.0

In [5]:
data_filtered['MerDate'] = pd.to_datetime(data_filtered['MerDate'])
data_filtered.set_index('MerDate', inplace=True)
aggregated_data = data_filtered.groupby([pd.Grouper(freq='15S'), 'Error/warnings']).size().unstack(fill_value=0)

/var/folders/pt/00xl4yzd64s0vbx9w9tzh94m0000gn/T/ipykernel_85212/2468120529.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['MerDate'] = pd.to_datetime(data_filtered['MerDate'])
/var/folders/pt/00xl4yzd64s0vbx9w9tzh94m0000gn/T/ipykernel_85212/2468120529.py:3: FutureWarning: 'S' is deprecated and will be removed in a future version, please use 's' instead.
  aggregated_data = data_filtered.groupby([pd.Grouper(freq='15S'), 'Error/warnings']).size().unstack(fill_value=0)


In [6]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Rename 'old_name' to 'new_name'
aggregated_data.rename(columns={'8_1': 'machine stops'}, inplace=True)

aggregated_data.head(10)


Error/warnings       21_100_1.0_0.0  21_100_255.0_0.0  21_100_27.0_0.0  \
MerDate                                                                  
2025-01-30 10:21:30               0                 0                0   
2025-01-30 10:21:45               0                 0                0   
2025-01-30 10:22:00               0                 0                0   
2025-01-30 10:22:15               0                 0                0   
2025-01-30 10:22:30               0                 0                0   
2025-01-30 10:22:45               0                 0                0   
2025-01-30 10:23:00               0                 0                0   
2025-01-30 10:23:15               0                 0                0   
2025-01-30 10:23:30               0                 0                0   
2025-01-30 10:23:45               0                 0                0   

Error/warnings       21_100_30.0_0.0  21_100_33.0_0.0  21_100_34.0_0.0  \
MerDate                                                                  
2025-01-30 10:21:30                0                0                0   
2025-01-30 10:21:45                0                0                0   
2025-01-30 10:22:00                0                0                0   
2025-01-30 10:22:15                0                0                0   
2025-01-30 10:22:30                0                0                0   
2025-01-30 10:22:45                0                0                0   
2025-01-30 10:23:00                0                0                0   
2025-01-30 10:23:15                0                0                0   
2025-01-30 10:23:30                0                0                0   
2025-01-30 10:23:45                0                0                0   

Error/warnings       21_100_7.0_0.0  21_101_1.0_0.0  21_101_2.0_0.0  \
MerDate                                                               
2025-01-30 10:21:30               0               0               0   
2025-01-30 10:21:45               0               0               0   
2025-01-30 10:22:00               0               0               0   
2025-01-30 10:22:15               0               0               0   
2025-01-30 10:22:30               0               0               0   
2025-01-30 10:22:45               0               0               0   
2025-01-30 10:23:00               0               0               0   
2025-01-30 10:23:15               0               0               0   
2025-01-30 10:23:30               0               0               0   
2025-01-30 10:23:45               0               0               0   

Error/warnings       21_101_3.0_0.0  21_102_47.0_0.0  21_102_50.0_0.0  \
MerDate                                                                 
2025-01-30 10:21:30               0                0                0   
2025-01-30 10:21:45               0                0                0   
2025-01-30 10:22:00               0                0                0   
2025-01-30 10:22:15               0                0                0   
2025-01-30 10:22:30               0                0                0   
2025-01-30 10:22:45               0                0                0   
2025-01-30 10:23:00               0                0                0   
2025-01-30 10:23:15               0                0                0   
2025-01-30 10:23:30               0                0                0   
2025-01-30 10:23:45               0                0                0   

Error/warnings       21_105_0.0_0.0  21_110_0.0_0.0  21_120_11.0_0.0  \
MerDate                                                                
2025-01-30 10:21:30               0               0                0   
2025-01-30 10:21:45               0               0                0   
2025-01-30 10:22:00               0               0                0   
2025-01-30 10:22:15               0               0                0   
2025-01-30 10:22:30               0               0         

In [7]:

# Assuming you already have your DataFrame loaded as aggregated_data
column_sums = aggregated_data.sum()

# Display the result
column_sums.sort_values(ascending=False)


Error/warnings
23_98_0.0_0.0        525026
22_98_0.0_0.0        524455
23_490_0.0_0.0        76955
23_458_0.0_0.0        65945
22_458_13.0_0.0       57559
22_490_12.0_0.0       50604
22_490_11.0_0.0       34067
23_96_0.0_0.0         29823
21_411_11.0_0.0       28625
22_453_12.0_0.0       25898
23_453_0.0_0.0        24816
21_414_11.0_0.0       21847
23_414_0.0_0.0        21440
23_411_0.0_0.0        20336
21_96_2.0_13.0        17267
21_456_11.0_0.0       14818
23_460_0.0_0.0        14464
21_452_0.0_0.0        14424
21_401_11.0_0.0       14203
23_456_0.0_0.0        13630
23_452_0.0_0.0        13181
22_461_0.0_0.0        11497
23_461_0.0_0.0        11295
21_460_22.0_41.0       8501
24_0_0.0_0.0           7069
22_458_12.0_0.0        6257
21_412_11.0_0.0        6021
21_460_17.0_0.0        5868
21_410_11.0_0.0        5769
23_412_0.0_0.0         5665
21_470_14.0_0.0        5648
8_1_96.0_18.0          5107
8_0_96.0_18.0          5107
21_415_11.0_0.0        4974
23_127_0.0_0.0         4680
23_47

- 490 pack final processing warning (an error occured at the end of line automation)
> meest voorkomend zijn case packer inhibit en take away conveyor full
- 98 rerouting activated for eject station (number of eggs at eject station has reached the maximum. The grader has stopped sending more eggs to this station/lane)
- 458 missing egg detected 
> meest voorkomend is Not stopped due to a setting
- 96 pressure control activated for eject station (number of eggs at eject station has reached the maximum. The grader needs to STOP to prevent cracking)
> meest voorkomend is egg buffer full lane 13
- 411 general safety failure error
> 
- 453 missing eggs warning (can be one or more in a pack)
- 414 cascade front door opened error (a safety switch is triggered) 
- 460 denester error



In [8]:
data_filtered['lane_stopping_machine'] = data_filtered.apply(lambda row: row['MerData3'] if row['MerNumber'] == 96 else pd.NA, axis=1)
data_filtered['lane_stopping_machine'].value_counts()


/var/folders/pt/00xl4yzd64s0vbx9w9tzh94m0000gn/T/ipykernel_85212/3648751589.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_filtered['lane_stopping_machine'] = data_filtered.apply(lambda row: row['MerData3'] if row['MerNumber'] == 96 else pd.NA, axis=1)


lane_stopping_machine
0.0     33829
13.0    21885
11.0     3876
12.0       61
Name: count, dtype: int64

- We want to see the duration of uptime of the machine. So that we can compare these as snapshots with eachother.
- There is every time another lane that caused the machine to stop

In [10]:
data_filtered[data_filtered['MerDate'] <= '2025-03-20 19:43:08']

KeyError: 'MerDate'

In [26]:
# Initialize variables
chunks = []
current_chunk = []
recording = False

# Iterate through the DataFrame
for index, row in data_filtered.iterrows():
    if row['MerType'] == 8 and row['MerNumber'] == 1:
        current_chunk = [row]
        recording = True
    elif recording:
        current_chunk.append(row)
        if row['MerType'] == 8 and row['MerNumber'] == 0:
            chunks.append(pd.DataFrame(current_chunk))
            current_chunk = []
            recording = False

print(f"Extracted and saved {len(chunks)} chunks.")


Extracted and saved 24081 chunks.


In [24]:
# Initialize variables
uptime_windows = []
start_time = None

# Iterate through the DataFrame to identify uptime windows
for index, row in data_filtered.iterrows():
    if row['MerType'] == 8 and row['MerNumber'] == 1:
        start_time = index
    elif row['MerType'] == 8 and row['MerNumber'] == 0 and start_time is not None:
        duration = index - start_time
        uptime_windows.append({
            'Start': start_time, 
            'Stop' : index,
            'Duration' : duration})
        start_time = None

# Display the results
for window in uptime_windows:
    print(f"Start: {window['Start']}, Stop: {window['Stop']}, Duration: {window['Duration']}")

Start: 2025-01-30 10:25:48, Stop: 2025-01-30 10:31:04, Duration: 0 days 00:05:16
Start: 2025-01-30 10:31:10, Stop: 2025-01-30 10:31:15, Duration: 0 days 00:00:05
Start: 2025-01-30 10:31:17, Stop: 2025-01-30 10:44:54, Duration: 0 days 00:13:37
Start: 2025-01-30 10:45:52, Stop: 2025-01-30 10:46:11, Duration: 0 days 00:00:19
Start: 2025-01-30 10:46:59, Stop: 2025-01-30 10:52:47, Duration: 0 days 00:05:48
Start: 2025-01-30 10:52:50, Stop: 2025-01-30 10:52:56, Duration: 0 days 00:00:06
Start: 2025-01-30 10:52:58, Stop: 2025-01-30 11:03:04, Duration: 0 days 00:10:06
Start: 2025-01-30 11:03:35, Stop: 2025-01-30 11:04:18, Duration: 0 days 00:00:43
Start: 2025-01-30 11:04:20, Stop: 2025-01-30 11:04:25, Duration: 0 days 00:00:05
Start: 2025-01-30 11:04:27, Stop: 2025-01-30 11:04:35, Duration: 0 days 00:00:08
Start: 2025-01-30 11:04:42, Stop: 2025-01-30 11:05:18, Duration: 0 days 00:00:36
Start: 2025-01-30 11:05:19, Stop: 2025-01-30 11:05:28, Duration: 0 days 00:00:09
Start: 2025-01-30 11:05:31, 